src/main/python/ttconv/scc/context.py

```diff
@@ -0,0 +1,479 @@
+#!/usr/bin/env python
+# -*- coding: UTF-8 -*-
+
+# Copyright (c) 2020, Sandflow Consulting LLC
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""SCC context"""
+
+from __future__ import annotations
+
+import copy
+from typing import Optional, Type, Tuple
+
+from ttconv.model import Div
+from ttconv.scc.caption_paragraph import SccCaptionParagraph
+from ttconv.scc.caption_style import SccCaptionStyle
+from ttconv.scc.codes.attribute_codes import SccAttributeCode
+from ttconv.scc.codes.control_codes import SccControlCode
+from ttconv.scc.codes.mid_row_codes import SccMidRowCode
+from ttconv.scc.codes.preambles_address_codes import SccPreambleAddressCode
+from ttconv.scc.config import SccReaderConfiguration, TextAlignment
+from ttconv.scc.word import SccWord
+from ttconv.style_properties import StyleProperties
+from ttconv.time_code import SmpteTimeCode
+
+ROLL_UP_BASE_ROW = 15
+
+
+class SccContext:
+  """SCC context for reader"""
+
+  def __init__(self, safe_area_x_offset: int, safe_area_y_offset: int, config: Optional[SccReaderConfiguration] = None):
+    # Caption paragraphs container
+    self.div: Optional[Div] = None
+
+    # Caption paragraphs counter
+    self.count: int = 0
+
+    # Screen safe area offsets
+    self.safe_area_x_offset = safe_area_x_offset
+    self.safe_area_y_offset = safe_area_y_offset
+
+    # Previously read SCC word value
+    self.previous_word: Optional[SccWord] = None
+    self.previous_word_type: Optional[Type] = None
+
+    # Buffered caption being built
+    self.buffered_caption = None
+    # Captions being displayed
+    self.active_caption: Optional[SccCaptionParagraph] = None
```

Comment:
Shouldn't we start with a non-empty active caption, esp. if the default style is paint-on?

Reply:
Shouldn't the begin time code be set when the first printable character is displayed, instead of when the caption is instantiated?

Reply:
The idea is to make the start of processing the same as the middle, i.e. there is always a front buffer and a back buffer active.

Maybe this is too much of a change for now. Let me know.

Reply:
> Shouldn't the begin time code be set when the first printable character is displayed, instead of when the caption is instantiated?

This is done for Paint-On only. The begin time must be set on EOC for Pop-On, and on CR for Roll-Up.

> The idea is to make the start of processing the same as the middle, i.e. there is always a front buffer and a back buffer active.

I understand the idea, but I think we cannot start writing into buffers without a minimum of context (a style at least), otherwise we're moving away from the spirit of the 608 specification...

But if it's a prerequisite, the change shouldn't be too big...

Reply:
608/708 are in fact streaming format (SCC added the timestamps) and a device could join the stream at any time and thus possibly in the middle of a caption.

What is the current behavior if we start parsing in the middle of a caption? Maybe we should add this to the unit tests?

Reply:
@palemieux I've changed the reader behavior so that it skips captions until a style is specified in 49d14db

Let me know what you think!